File: research/experiments/sweep/runs/opt-medium-go-sonnet45/finite_diff_test.go
package optimization

import (
	"math"
	"testing"
)

// Test functions with known analytic gradients

func beale(x []float64) float64 {
	a := 1.5 - x[0] + x[0]*x[1]
	b := 2.25 - x[0] + x[0]*x[1]*x[1]
	c := 2.625 - x[0] + x[0]*x[1]*x[1]*x[1]
	return a*a + b*b + c*c
}

func bealeGrad(x []float64) []float64 {
	a := 1.5 - x[0] + x[0]*x[1]
	b := 2.25 - x[0] + x[0]*x[1]*x[1]
	c := 2.625 - x[0] + x[0]*x[1]*x[1]*x[1]

	dfdx0 := 2*a*(-1+x[1]) + 2*b*(-1+x[1]*x[1]) + 2*c*(-1+x[1]*x[1]*x[1])
	dfdx1 := 2*a*x[0] + 2*b*x[0]*2*x[1] + 2*c*x[0]*3*x[1]*x[1]

	return []float64{dfdx0, dfdx1}
}

func TestForwardDiffGradient(t *testing.T) {
	tests := []struct {
		name      string
		f         func([]float64) float64
		analyticG func([]float64) []float64
		x         []float64
		tol       float64
	}{
		{
			name:      "sphere at [3,4]",
			f:         sphere,
			analyticG: sphereGrad,
			x:         []float64{3, 4},
			tol:       1e-7,
		},
		{
			name:      "sphere at origin",
			f:         sphere,
			analyticG: sphereGrad,
			x:         []float64{0, 0},
			tol:       1e-7,
		},
		{
			name:      "rosenbrock at [-1.2,1]",
			f:         rosenbrock,
			analyticG: rosenbrockGrad,
			x:         []float64{-1.2, 1.0},
			tol:       1e-4,
		},
		{
			name:      "beale at [1,1]",
			f:         beale,
			analyticG: bealeGrad,
			x:         []float64{1, 1},
			tol:       1e-5,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			numericGrad := ForwardDiffGradient(tt.f, tt.x)
			analyticGrad := tt.analyticG(tt.x)

			for i := range numericGrad {
				diff := math.Abs(numericGrad[i] - analyticGrad[i])
				if diff > tt.tol {
					t.Errorf("Component %d: numeric=%v, analytic=%v, diff=%v > tol=%v",
						i, numericGrad[i], analyticGrad[i], diff, tt.tol)
				}
			}
		})
	}
}

func TestCentralDiffGradient(t *testing.T) {
	tests := []struct {
		name      string
		f         func([]float64) float64
		analyticG func([]float64) []float64
		x         []float64
		tol       float64
	}{
		{
			name:      "sphere at [3,4]",
			f:         sphere,
			analyticG: sphereGrad,
			x:         []float64{3, 4},
			tol:       1e-10,
		},
		{
			name:      "sphere at origin",
			f:         sphere,
			analyticG: sphereGrad,
			x:         []float64{0, 0},
			tol:       1e-10,
		},
		{
			name:      "rosenbrock at [-1.2,1]",
			f:         rosenbrock,
			analyticG: rosenbrockGrad,
			x:         []float64{-1.2, 1.0},
			tol:       1e-7,
		},
		{
			name:      "beale at [1,1]",
			f:         beale,
			analyticG: bealeGrad,
			x:         []float64{1, 1},
			tol:       1e-8,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			numericGrad := CentralDiffGradient(tt.f, tt.x)
			analyticGrad := tt.analyticG(tt.x)

			for i := range numericGrad {
				diff := math.Abs(numericGrad[i] - analyticGrad[i])
				if diff > tt.tol {
					t.Errorf("Component %d: numeric=%v, analytic=%v, diff=%v > tol=%v",
						i, numericGrad[i], analyticGrad[i], diff, tt.tol)
				}
			}
		})
	}
}

func TestMakeGradient(t *testing.T) {
	x := []float64{3, 4}

	t.Run("default to forward", func(t *testing.T) {
		gradFunc := MakeGradient(sphere, "")
		numericGrad := gradFunc(x)
		forwardGrad := ForwardDiffGradient(sphere, x)

		if !vecEqual(numericGrad, forwardGrad) {
			t.Errorf("MakeGradient with empty method should use forward differences")
		}
	})

	t.Run("explicit forward", func(t *testing.T) {
		gradFunc := MakeGradient(sphere, "forward")
		numericGrad := gradFunc(x)
		forwardGrad := ForwardDiffGradient(sphere, x)

		if !vecEqual(numericGrad, forwardGrad) {
			t.Errorf("MakeGradient with 'forward' method should use forward differences")
		}
	})

	t.Run("central", func(t *testing.T) {
		gradFunc := MakeGradient(sphere, "central")
		numericGrad := gradFunc(x)
		centralGrad := CentralDiffGradient(sphere, x)

		if !vecEqual(numericGrad, centralGrad) {
			t.Errorf("MakeGradient with 'central' method should use central differences")
		}
	})
}

func TestFiniteDiffImmutability(t *testing.T) {
	x := []float64{3, 4}
	xCopy := Clone(x)

	ForwardDiffGradient(sphere, x)
	if !vecEqual(x, xCopy) {
		t.Errorf("ForwardDiffGradient modified input vector")
	}

	CentralDiffGradient(sphere, x)
	if !vecEqual(x, xCopy) {
		t.Errorf("CentralDiffGradient modified input vector")
	}
}

File: research/experiments/sweep/runs/opt-medium-cpp-sonnet45/tests/finite_diff_test.cpp
#include "doctest.h"
#include "finite_diff.h"
#include <cmath>

using namespace optimization;

// Test functions with analytic gradients
static double sphere(const Vector& x) {
    double sum = 0.0;
    for (double xi : x) {
        sum += xi * xi;
    }
    return sum;
}

static Vector sphereGrad(const Vector& x) {
    Vector grad(x.size());
    for (size_t i = 0; i < x.size(); ++i) {
        grad[i] = 2.0 * x[i];
    }
    return grad;
}

static double rosenbrock(const Vector& x) {
    double a = 1.0 - x[0];
    double b = x[1] - x[0] * x[0];
    return a * a + 100.0 * b * b;
}

static Vector rosenbrockGrad(const Vector& x) {
    double a = 1.0 - x[0];
    double b = x[1] - x[0] * x[0];
    return {
        -2.0 * a - 400.0 * b * x[0],
        200.0 * b
    };
}

static double beale(const Vector& x) {
    double t1 = 1.5 - x[0] + x[0] * x[1];
    double t2 = 2.25 - x[0] + x[0] * x[1] * x[1];
    double t3 = 2.625 - x[0] + x[0] * x[1] * x[1] * x[1];
    return t1 * t1 + t2 * t2 + t3 * t3;
}

static Vector bealeGrad(const Vector& x) {
    double t1 = 1.5 - x[0] + x[0] * x[1];
    double t2 = 2.25 - x[0] + x[0] * x[1] * x[1];
    double t3 = 2.625 - x[0] + x[0] * x[1] * x[1] * x[1];

    double dx = 2.0 * t1 * (-1.0 + x[1]) +
                2.0 * t2 * (-1.0 + x[1] * x[1]) +
                2.0 * t3 * (-1.0 + x[1] * x[1] * x[1]);

    double dy = 2.0 * t1 * x[0] +
                2.0 * t2 * 2.0 * x[0] * x[1] +
                2.0 * t3 * 3.0 * x[0] * x[1] * x[1];

    return {dx, dy};
}

TEST_CASE("finite-diff: forward difference on sphere at [3,4]") {
    Vector x = {3, 4};
    Vector grad = forwardDiffGradient(sphere, x);
    Vector expected = sphereGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-7));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-7));
}

TEST_CASE("finite-diff: forward difference on sphere at [0,0]") {
    Vector x = {0, 0};
    Vector grad = forwardDiffGradient(sphere, x);
    Vector expected = sphereGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-7));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-7));
}

TEST_CASE("finite-diff: forward difference on Rosenbrock") {
    Vector x = {-1.2, 1.0};
    Vector grad = forwardDiffGradient(rosenbrock, x);
    Vector expected = rosenbrockGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-4));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-4));
}

TEST_CASE("finite-diff: forward difference on Beale") {
    Vector x = {1, 1};
    Vector grad = forwardDiffGradient(beale, x);
    Vector expected = bealeGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-5));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-5));
}

TEST_CASE("finite-diff: central difference on sphere at [3,4]") {
    Vector x = {3, 4};
    Vector grad = centralDiffGradient(sphere, x);
    Vector expected = sphereGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-10));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-10));
}

TEST_CASE("finite-diff: central difference on sphere at [0,0]") {
    Vector x = {0, 0};
    Vector grad = centralDiffGradient(sphere, x);
    Vector expected = sphereGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-10));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-10));
}

TEST_CASE("finite-diff: central difference on Rosenbrock") {
    Vector x = {-1.2, 1.0};
    Vector grad = centralDiffGradient(rosenbrock, x);
    Vector expected = rosenbrockGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-7));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-7));
}

TEST_CASE("finite-diff: central difference on Beale") {
    Vector x = {1, 1};
    Vector grad = centralDiffGradient(beale, x);
    Vector expected = bealeGrad(x);

    CHECK(grad[0] == doctest::Approx(expected[0]).epsilon(1e-8));
    CHECK(grad[1] == doctest::Approx(expected[1]).epsilon(1e-8));
}

TEST_CASE("finite-diff: makeGradient factory - forward") {
    auto gradFunc = makeGradient(sphere);
    Vector x = {3, 4};
    Vector grad = gradFunc(x);
    Vector expected = forwardDiffGradient(sphere, x);

    CHECK(grad[0] == doctest::Approx(expected[0]));
    CHECK(grad[1] == doctest::Approx(expected[1]));
}

TEST_CASE("finite-diff: makeGradient factory - central") {
    auto gradFunc = makeGradient(sphere, "central");
    Vector x = {3, 4};
    Vector grad = gradFunc(x);
    Vector expected = centralDiffGradient(sphere, x);

    CHECK(grad[0] == doctest::Approx(expected[0]));
    CHECK(grad[1] == doctest::Approx(expected[1]));
}
